src/AbstractDatabase.php

```diff
@@ -0,0 +1,285 @@
+
+    /**
+     * The placeholder to use for the status column name in the version log table.
+     *
+     * @since [*next-version*]
+     */
+    const PLACEHOLDER_LOG_STATUS_COLUMN = '{lt_status}';
+
+    /**
+     * The version status used to indicate completed migrations.
+     *
+     * @since [*next-version*]
+     */
+    const STATUS_COMPLETE = 'complete';
+
+    /**
+     * The version status used to indicated partial migrations.
+     *
+     * @since [*next-version*]
+     */
+    const STATUS_PARTIAL = 'partial';
+
+    /**
+     * The version status used to indicated an unknown migration state, typically used for the base version.
+     *
+     * @since [*next-version*]
+     */
+    const STATUS_UNKNOWN = 'unknown';
+
+    /**
+     * Formats a query string.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $sql  The SQL query string to format.
+     * @param array  $args Optional arguments to interpolate into the SQL string.
+     *
+     * @return string The formatted SQL query.
+     */
+    protected function _formatSql($sql, array $args = [])
+    {
+        $sql = $this->_normalizeString($sql);
+        $sql = vsprintf($sql, $args);
+        $sql = str_replace(static::PLACEHOLDER_DATABASE, $this->_getDatabaseName(), $sql);
+        $sql = str_replace(static::PLACEHOLDER_LOG_TABLE, $this->_getLogTableName(), $sql);
+        $sql = str_replace(static::PLACEHOLDER_LOG_VERSION_COLUMN, $this->_getLogTableVersionColumn(), $sql);
+        $sql = str_replace(static::PLACEHOLDER_LOG_STATUS_COLUMN, $this->_getLogTableStatusColumn(), $sql);
+
+        return $sql;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getVersion()
+    {
+        $result     = [];
+        $versionCol = $this->_getLogTableVersionColumn();
+        $statusCol  = $this->_getLogTableStatusColumn();
+
+        try {
+            $result[$versionCol] = $this->getDbDriver()->getScalar(
+                $this->_formatSql(
+                    'SELECT %1$s FROM %2$s',
+                    [
+                        static::PLACEHOLDER_LOG_VERSION_COLUMN,
+                        static::PLACEHOLDER_LOG_TABLE,
+                    ]
+                )
+            );
+        } catch (\Exception $ex) {
+            throw new DatabaseNotVersionedException(
+                'This database does not have a migration version. Please use "migrate reset" or "migrate install" to create one.'
+            );
+        }
+
+        try {
+            $result[$statusCol] = $this->getDbDriver()->getScalar(
+                $this->_formatSql(
+                    'SELECT %1$s FROM %2$s',
+                    [
+                        static::PLACEHOLDER_LOG_STATUS_COLUMN,
+                        static::PLACEHOLDER_LOG_TABLE,
+                    ]
+                )
+            );
+        } catch (\Exception $ex) {
+            throw new OldVersionSchemaException(
```

Comment:
It's what the parent class throws, but yes it makes more sense to throw `DatabaseNotVersionedException`. I'll change it.

Reply:
I think you misunderstood. You are throwing this in case of *any* exception. For example, down the stack there will be a call to `_normalizeString()`, which is allowed to throw `InvalidArgumentException`. Does this mean that "this database does not have a migration version"?

Reply:
What would you do? 😆 This is what the parent class did and I simply copied it. There are no documented throws for the DB adapter and driver. The only way to know what exceptions can be thrown is to look at the implementations - of which there already are many.

Perhaps the most sensible thing to do here is to simply not catch.

Reply:
Hmm, good point. Maybe, use a custom (extended) driver, which narrows what is thrown?

Reply:
Maybe. But that would mean we're only catering for that one driver. Is that ok?

Reply:
Shiet. Ok, do what you think is best.

Reply:
I couldn't help myself so I did some digging.

It seems that it would be best to catch a `PDOException`. The original byjg implementation was attempt to catch exceptions thrown when the query that checks for the DB version fails, and thus it deduces that the DB is not versioned.

Catching `PDOException` should be broad enough to cover such exceptions, while also not being too broad that it would catch non-query related exceptions.